#+title: Webtables


* MO Tables
- Organism
  - [X] OrganismAbbreviationBlast_ix
  - [X] OrganismAbbreviationBlast
    - move to KEEP
  - [X] OrganismAbbreviation_ix
  - [X] OrganismAbbreviation
    - updated the abbreviation field to name_for_filenames
  - [X] OrganismSelectTaxonRank_ix
  - [X] OrganismSelectTaxonRank
  - [X] Taxonomy_ix
  - [X] Taxonomy
  - [X] TaxonSpecies_ix
  - [X] TaxonSpecies
- Genomic Sequence
  - [X] GenomicSequenceId_ix
  - [X] GenomicSequenceId
  - [X] GenomicSequenceSequence_ix
  - [X] GenomicSequenceSequence
  - [X] SequencePieceClosure
  - [X] GenomicSeqAttributes
  - [s] SequenceEnzymeClass
    - Temp remove this and eventually Move to ComparativeGenomics because it depends on the OrthoMCL Derived EC Numbers
    
- Transcript / Protein
  - [X] SignalPeptideDomains_ix
  - [X] SignalPeptideDomains
  - [X] TransmembraneDomains_ix
  - [X] TransmembraneDomains
  - [X] PdbSimilarity_ix
  - [X] PdbSimilarity
  - [X] ProteinSequence_ix
  - [X] ProteinSequence
  - [ ] ProteinAttributes_ix
    - [ ] Remove Derived EC Numbers
    - [ ] New Tuning table(s) for EC Derived ECs per protein (gene and transcript)
  - [X] ProteinAttributes
  - [ ] TranscriptAttributes_ix
  - [ ] TranscriptAttributes
    - no longer has products column because that is done by TM
    - no longer has derived ec numbers (move to comparative genomics)
  - [X] CodingSequence_ix
  - [X] CodingSequence
  - [X] IntronUtrCoords_ix
  - [X] IntronUtrCoords
  - [X] TranscriptCenDistance_ix
  - [X] TranscriptCenDistance
  - [ ] TranscriptPathway_ix
  - [ ] TranscriptPathway
    - This may need to move to comparative genomics because we need the OrthoDerived EC mappings
  - [X] TranscriptSequence_ix
  - [X] TranscriptSequence
  - [X] ChIPchipTranscript_ix
  - [X] ChIPchipTranscript

- Gene
  - [X] GeneId_ix
  - [X] GeneId
  - [X] GeneAttributes_ix
  - [X] GeneAttributes
    - remove gene_product and remove orthomclname
  - [X] GeneCopyNumbers_ix
  - [X] GeneCopyNumbers
  - [X] GeneGoTable_ix
  - [X] GeneGoTable
  - [X] GeneGoTerms_ix
  - [X] GeneGoTerms
  - [X] GeneLocations_ix
  - [X] GeneLocations
  - [X] GeneModelDump_ix
  - [X] GeneModelDump
  - [X] GeneSummaryFilter_ix
  - [X] GeneSummaryFilter
  - [X] TFBSGene_ix
  - [X] TFBSGene
    - removed aef.*
  - [ ] PathwayNodeGene_ix
    - This may need to move to comparative genomics because we need the OrthoDerived EC mappings
  - [ ] PathwayNodeGene
    - This may need to move to comparative genomics because we need the OrthoDerived EC mappings
  - [ ] PathwaysGeneTable_ix
    - This may need to move to comparative genomics because we need the OrthoDerived EC mappings
  - [ ] PathwaysGeneTable
    - This may need to move to comparative genomics because we need the OrthoDerived EC mappings
  - [X] GoTermSummary_ix
  - [X] GoTermSummary
  - [X] EqtlSpan_ix
  - [X] EqtlSpan

- EST
  - [X] EstAttributes_ix
    - move to comparative
    - join to apidb.organism and filter by "is_reference_strain"
  - [X] EstAttributes
    - move to comparative
    - join to apidb.organism and filter by "is_reference_strain"
  - [X] EstSequence_ix
    - move to comparative
    - join to apidb.organism and filter by "is_reference_strain"
  - [X] EstSequence
    - move to comparative
    - join to apidb.organism and filter by "is_reference_strain"
  - [X] EstAlignmentGeneSummary_ix
  - [X] EstAlignmentGeneSummary

- Dataset / Other
  - [X] DatasetExampleSourceId_ix
  - [X] DatasetExampleSourceId
    - NOTE:  this depends on Profiles
  - [X] PANExtDBRls
  - [X] PANIO
  - [X] PANIO_ix

  - [ ] +ProfileType_ix+
  - [ ] +ProfileType+
  - [ ] +Profile_ix+
  - [ ] +Profile+
  - [ ] +ProfileSamples_ix+
  - [ ] +ProfileSamples+

  - [X] RnaSeqStats_ix
  - [X] RnaSeqStats
  - [X] OrganismAttributes_ix
  - [X] OrganismAttributes
    - removed ESTs and SNPs
  - [X] ChrCopyNumbers_ix
  - [X] ChrCopyNumbers

- Junctions (Kathryn)
  - [ ] IntronSupportLevel_ix
  - [ ] IntronSupportLevel
  - [ ] GeneIntJuncStats_ix
  - [ ] GeneIntJuncStats
  - [ ] GeneIntronJunction_ix
  - [ ] GeneIntronJunction
  - [ ] NameMappingGIJ_ix
  - [ ] NameMappingGIJ
    - needs to be in TM, depends on dataset presenters
  - [ ] GeneMaxIntronGIJ_ix
    - should be aux table and dropped
  - [ ] GeneMaxIntronGIJ
    - should be aux table and dropped

- Comparative
  - [ ] ProteinGroup (RENMAED from SEQUENCEATTRIBUTES)
    - add orthomcl derived ec numbers here
    - add gene_id
    - add transcript_id
    - ENSURE this has a row for every protein!
  - [ ] ProteinGroupDomainAssignment (RENAMED from DOMAINASSIGNMENT)

  - [ ] GroupDomainDescriptions (RENAMED from GROUPDOMAINATTRIBUTE )
  - [ ] proteinGroupEnyzmeClass (renamed from sequenceenzymeclass)
  - [ ] AlphaFoldGenes
  - [ ] GroupPhylogeneticProfile
  - [ ] OrthologousTranscripts
  - [ ] PhyleticPattern
    - no longer uses dots.sequencegroup and dots.sequencesequencegroup

- Global
  - [X] CompoundAttributes
  - [X] CompoundId
  - [X] CompoundProperties
  - [ ] CompoundTypeAheads
    - double check again after database is rebuilt.  may be ok
  - [X] OntologyLevels
  - [X] PathwayAttributes
  - [X] PathwayCompounds
  - [X] PathwayNodes
  - [X] PathwayReactions

- KEEP Tuning table
  - GeneProduct (CHECK)
  - TranscriptProduct (NEW)
  - ProteinProduct (NEW??)
